examples/workshop-fomo-generalized/index.rsh

```diff
@@ -0,0 +1,68 @@
+'reach 0.1';
+
+// FOMO Workshop generalized to last N winners
+const NUM_OF_WINNERS = 3;
+
+const CommonInterface = {
+  showOutcome: Fun([Array(Address, NUM_OF_WINNERS)], Null),
+};
+
+const FunderInterface = {
+  ...CommonInterface,
+  getParams: Fun([], Object({
+    deadline: UInt, // relative deadline
+    ticketPrice: UInt,
+  })),
+};
+
+const BuyerInterface = {
+  ...CommonInterface,
+  shouldBuyTicket: Fun([UInt], Bool),
+  showPurchase: Fun([Address], Null),
+};
+
+export const main = Reach.App(
+  { },
+  [
+    ['Funder', FunderInterface], ['class', 'Buyer', BuyerInterface],
+  ],
+  (Funder, Buyer) => {
+    const showOutcome = (winners) =>
+      each([Funder, Buyer], () => interact.showOutcome(winners));
+
+    Funder.only(() => {
+      const { ticketPrice, deadline } = declassify(interact.getParams()); });
+    Funder.publish(ticketPrice, deadline);
+
+    const initialWinners = Array.replicate(NUM_OF_WINNERS, Funder);
+
+    // Until deadline, allow buyers to buy ticket
+    const [ keepGoing, winners, ticketsSold ] =
+      parallel_reduce([ true, initialWinners, 0 ])
+        .invariant(balance() == ticketsSold * ticketPrice)
+        .while(keepGoing)
+        .case(
+          Buyer,
+          (() => ({
+            when: declassify(interact.shouldBuyTicket(ticketPrice)) })),
+          (() => ticketPrice),
+          () => {
+            const buyer = this;
+            Buyer.only(() => interact.showPurchase(buyer));
+            const idx = ticketsSold % NUM_OF_WINNERS;
+            const newWinners =
+              Array.set(winners, idx, buyer);
+            return [ true, newWinners, ticketsSold + 1 ]; })
+        .timeout(deadline, () => {
+          race(Buyer, Funder).publish();
+          return [ false, winners, ticketsSold ]; });
+
+    transfer(balance() % NUM_OF_WINNERS).to(Funder);
```

Comment:
This is just to get rid of the excess balance to satisfy the 0 balance theorem, right? If there are 10 winners, I'm pretty sure "send up to 10 WEI to the Funder" would just end up costing _more_ in gas. Not sure what the right move is here, but this is probably fine for now. (Maybe add a larger "funder's fee" to justify the transfer?)

Reply:
That's a good point. For the sake of having simple, beautiful code in the workshop, I think it might not be worth tweaking with concern of gas costs, as it will introduce more complexity not related to the game mechanism. I think that should be an exercise left up to the reader (: